// homey-wifidriver/lib/driver.js
'use strict';

const Homey = require('homey');

const OAuth2Account = require('./OAuth2/OAuth2Account');
const OAuth2ClientManager = require('./OAuth2/OAuth2ClientManager');

class HomeyWifiDriver extends Homey.Driver {

	/**
	 * This method needs to be called from a driver.onInit(), it will create an OAuth2 client
	 * and override the default onPair() method to handle the OAuth2 Flow.
	 * @param {Object} options
	 * @param {Object} options.oauth2ClientConfig - OAuth2 client configuration object
	 * @param {string} options.oauth2ClientConfig.url - OAuth2 authorization url
	 * @param {string} options.oauth2ClientConfig.tokenEndpoint - OAuth2 token endpoint
	 * @param {string} options.oauth2ClientConfig.key - OAuth2 client key
	 * @param {string} options.oauth2ClientConfig.secret - OAuth2 client secret
	 * @param {string} options.oauth2ClientConfig.allowMultipleAccounts - Specifies whether multiple accounts per
	 * client are allowed
	 */
	onInit(options) {
		options = options || {};

		// Check if OAuth2 client config is provided
		if (options.oauth2ClientConfig) {

			this.oauth2ClientConfig = options.oauth2ClientConfig;
			this.oauth2ClientConfig.log = this.log.bind(this);

			// If there is no client manager present create one
			if (!(Homey.app.OAuth2ClientManager instanceof OAuth2ClientManager))
				Homey.app.OAuth2ClientManager = new OAuth2ClientManager({ log: this.log.bind(this) });

			// Create a new client or return existing one depending on configuration
			this.oauth2Client = Homey.app.OAuth2ClientManager.createClient(this.oauth2ClientConfig);

			this.log('onInit() -> initialized OAuth2 client');

			// Override the onPair method
			this.onPair = this.customOAuth2OnPair.bind(this);

			this.log('onInit() -> redirect onPair() to OAuth2 pairing flow');
		}
	}

	/**
	 * This method will be called when the user starts pairing a new device,
	 * it executes a OAuth2 flow to retrieve devices linked to the users account.
	 * @param socket
	 */
	customOAuth2OnPair(socket) {

		// Create OAuth2 account from config
		const oauth2Account = new OAuth2Account(this.oauth2ClientConfig);
		const oauth2Url = oauth2Account.getOAuth2Url();

		new Homey.CloudOAuth2Callback(oauth2Url)
			.once('url', url => {
				this.log('customOAuth2OnPair() -> retrieved authentication url');
				socket.emit('url', url);
			})
			.once('code', code => {
				this.log('customOAuth2OnPair() -> retrieved authentication code');

				// Fetch access tokens
				oauth2Account.getAccessTokens(code)
					.then(() => {
						this.log('customOAuth2OnPair() -> succesfully retrieved access tokens');
						return socket.emit('authorized');
					})
					.catch(err => {
						this.error(err.stack);
						socket.emit('error', err.message);
					});
			})
			.generate();

		// Listen for list_devices request
		socket.on('list_devices', (data, callback) => {
			data = data || {};

			// Put OAuth2 account in data object
			data.oauth2Account = oauth2Account;

			this.log('customOAuth2OnPair() -> fetch list devices');

			// Call driver's OAuth2 list devices method to ask for devices list
			this.onPairOAuth2ListDevices(data)
				.then(devices => callback(null, devices || []))
				.catch(err => {
					this.error(err.stack);
					return callback(err);
				});
		});

		// Destroy oauth2Account when pairing is finished, device.onInit() will create permanent account
		socket.on('disconnect', () => oauth2Account.destroy());
	}
}

module.exports = HomeyWifiDriver;
