// node-toon/node-toon.js
'use strict';

var request = require("request");
var EventEmitter = require('events').EventEmitter;

class Toon extends EventEmitter {

	/**
	 * Toon constructor, provide the API key and secret
	 * @constructor
	 */
	constructor(key, secret) {
		super();

		// Store key and secret for authorization later on
		this.key = key;
		this.secret = secret;

		// Defaults
		this.target_temperature = undefined;
		this.measure_temperature = undefined;

		// Create fields for the access tokens
		this.access_token = new Buffer(this.key + ":" + this.secret).toString('base64');
		this.refresh_token = null;

		// Start listening for incoming events // TODO when pushEvent works
		// this._listenForEvents();
	}

	/**
	 * Queries the Toon API for the display status
	 * @param callback Returns with two parameters, err and data
	 */
	getStatus(callback) {

		console.log("node-toon: get status");

		this._baseGET("status", (err, result) => {
			if (!err && result) {
				if (result && result.thermostatInfo) {

					// Store updated values
					this.measure_temperature = Math.round((result.thermostatInfo.currentTemp / 100) * 10) / 10 || this.measure_temperature;
					this.target_temperature = Math.round((result.thermostatInfo.currentSetpoint / 100) * 10) / 10 || this.target_temperature;
				}

				if (callback) callback(err, result.thermostatInfo);
			}
			else {

				console.log("node-toon: no new data available");

				if (callback) callback(err, false);
			}
		})
	}

	/**
	 * PUTs to the Toon API to set a new target temperature
	 * @param data Object holding a temperature attribute of type integer
	 * @param callback Returns with two parameters, err and success
	 */
	setTargetTemperature(temperature, callback) {

		if (temperature) temperature = temperature * 100;

		console.log("node-toon: set target temperature to " + temperature);

		this._basePUT("temperature", { "value": temperature, "scale": "CELSIUS" }, (err, result) => {

			if (result) console.log("node-toon: success setting temperature to " + temperature);
			if (err) console.log("node-toon: failed to set temperature to " + temperature);

			if (callback) callback(err, result)
		})
	}

	/**
	 * Returns current target temperature
	 */
	getTargetTemperature(callback) {
		callback(null, this.target_temperature);
	}

	/**
	 * Returns current measured temperature
	 */
	getMeasureTemperature(callback) {
		callback(null, this.measure_temperature);
	}

	/**
	 * Queries the Toon API for the temperature programs
	 * @param callback Returns with two parameters, err and data
	 */
	getTemperaturePrograms(callback) {
		this._baseGET("temperature/programs", (err, result) => {
			if (callback) callback(err, result)
		})
	}

	/**
	 * Queries the Toon API for the temperature programs
	 * @param callback Returns with two parameters, err and data
	 */
	getConsumptionElectricity(callback) {
		this._baseGET("consumption/electricity/data*", (err, result) => {
			if (callback) callback(err, result)
		})
	}

	/**
	 * Queries the Toon API for the temperature programs
	 * @param callback Returns with two parameters, err and data
	 */
	getConsumptionGas(callback) {
		this._baseGET("consumption/gas/data*", (err, result) => {
			if (callback) callback(err, result)
		})
	}

	/**
	 * Fetches all agreements from the API, if there are more
	 * than one, the user may choose one
	 * @param callback
	 */
	getAgreements(callback) {

		console.log("node-toon: get agreements: " + this.access_token);

		this._baseGET("agreements", (err, result) => {
			if (!err && result) {

				console.log(`node-toon: got ${result.length} agreements`);

				this.recentAgreements = result;
			} else console.log("node-toon: failed to get agreements: " + err);

			if (callback) callback(err, result);
		});
	}

	/**
	 * Selects an agreement (device?) and registers it to this
	 * toon object, this is a connection to the device
	 * @param agreementId
	 * @param callback
	 */
	setAgreement(agreementId, callback) {

		// Either use parameter agreement
		let agreement;
		if (agreementId != null) {
			agreement = {
				agreementId: agreementId
			};
		}
		else {
			// Or use the most recent agreement
			agreement = this.recentAgreements[0];
		}

		console.log("node-toon: set agreement");

		// Make the request to set agreement
		this._basePOST("agreements", agreement, (err, result) => {

			if (result) console.log("node-toon: successful post of agreement");
			if (err) console.log("node-toon: failed to post agreement");

			// Check for updated target temperature
			this.getStatus(() => {

				// Agreements done
				if (callback) callback(err, result);
			});
		});
	}

	/**
	 * Fetches an access token from the Toon API using the
	 * athom callback service (redirect uri)
	 * @param code
	 * @param redirect_uri
	 * @param callback
	 * @private
	 */
	getAccessTokens(code, redirect_uri, callback) {
		let key = this.key;
		let secret = this.secret;

		// Request access_token
		request.post({
			url: "https://api.toonapi.com/token",
			rejectUnauthorized: false,
			form: {
				"grant_type": "authorization_code",
				"client_id": key,
				"client_secret": secret,
				"redirect_uri": redirect_uri,
				"code": code
			}
		}, (err, res, body) => {

			console.log("node-toon: fetched new access tokens");

			// Store new tokens
			this.access_token = JSON.parse(body).access_token;
			this.refresh_token = JSON.parse(body).refresh_token;

			// Callback new tokens
			callback(err, {
				access_token: JSON.parse(body).access_token,
				refresh_token: JSON.parse(body).refresh_token
			});
		});
	}

	/**
	 * Uses the refresh token to fetch a new access token,
	 * stores all new tokens internally
	 * @param callback
	 * @private
	 */
	refreshAccessToken(callback) {
		let refresh_token = this.refresh_token;

		// Refresh access_token
		request.post({
			url: "https://api.toonapi.com/token",
			rejectUnauthorized: false,
			headers: {
				"Authorization": "Bearer " + new Buffer(this.key + ":" + this.secret).toString('base64')
			},
			form: {
				"grant_type": "refresh_token",
				"refresh_token": refresh_token
			}
		}, (err, res, body) => {

			console.log("node-toon: fetched new (refresh) access tokens");

			// Store new tokens
			this.access_token = JSON.parse(body).access_token || this.access_token;
			this.refresh_token = JSON.parse(body).refresh_token || this.refresh_token;

			// Callback new tokens
			if (callback) callback(err, {
				access_token: JSON.parse(body).access_token,
				refresh_token: JSON.parse(body).refresh_token
			});
		});
	}

	/**
	 * Destroy this toon instance,
	 * and make sure it stops polling.
	 */
	destroy() {

		// Configure the request
		var options = {
			url: 'https://api.toonapi.com/toon/api/v1/pushEvent?' + Homey.env.TOON_KEY,
			method: 'DELETE',
			headers: {
				"Authorization": "Bearer " + this.access_token,
			}
		};

		console.log("destroy toon instance")
		// Start the request
		request(options, (error, response, body) => {
			console.log("destroy response: " + response.statusCode);

			if (!error && response.statusCode == 200) {
				var result = (body) ? JSON.parse(body) : true;
				console.log(error)
				console.log(result)
			}
			else {
				console.log(error)
			}
		});
	}

	/**
	 * Convenience method that provides a basic PUT
	 * to the Toon API
	 * @param command Desired command to be PUT
	 * @param body Data to be updated
	 * @param callback Returns with two parameters, err and success
	 * @private
	 */
	_basePUT(command, body, callback) {

		// Configure the request
		var options = {
			url: 'https://api.toonapi.com/toon/api/v1/' + command,
			method: 'PUT',
			"rejectUnauthorized": false,
			headers: {
				"authorization": "Bearer " + this.access_token,
				"Accept": "application/json"
			},
			json: body
		};

		// Perform the request
		this._performRequest(options, callback);
	}

	/**
	 * Convenience method that provides a basic GET
	 * to the Toon API
	 * @param command Desired command to be GET
	 * @param callback Returns with two parameters, err and success
	 * @private
	 */
	_baseGET(command, callback) {

		// Configure the request
		var options = {
			url: 'https://api.toonapi.com/toon/api/v1/' + command,
			method: 'GET',
			"rejectUnauthorized": false,
			headers: {
				"Authorization": "Bearer " + this.access_token,
				"Accept": "application/json"
			}
		};

		// Perform the request
		this._performRequest(options, callback);
	}

	/**
	 * Convenience method that provides a basic GET
	 * to the Toon API
	 * @param command Desired command to be GET
	 * @param callback Returns with two parameters, err and success
	 * @private
	 */
	_basePOST(command, data, callback) {

		// Perform request
		request({
			url: 'https://api.toonapi.com/toon/api/v1/' + command,
			method: "POST",
			headers: {
				"Authorization": "Bearer " + this.access_token,
				"Content-Type": "application/json"
			},
			json: data
		}, (err, result) => {
			if (callback) callback(err, result);
		});
	}

	/**
	 * Convenience method that perfoms a request to
	 * the Toon api, using the options provided in the
	 * parameter options
	 * @param options Request options
	 * @param callback Returns with two parameters, err and result
	 * @private
	 */
	_performRequest(options, callback) {

		// Start the request
		request(options, (error, response, body) => {
			if (!error && response.statusCode == 200) {
				var result = (body) ? JSON.parse(body) : true;
				if (callback) callback(error, result);
			}
			else {
				if (callback) callback((error || response.statusCode), response);
			}
		});
	}
}

module.exports = Toon;
